File: schemaorggo/invoice.go
package schemaorggo

import "encoding/json"

// Invoice see : https://schema.org/Invoice
type Invoice struct {
	Intangible

	typeContext

	// AccountId see : https://schema.org/accountId
	// The identifier for the account the payment will be applied to.
	// types : Text
	AccountId []string `json:"accountId,omitempty"`

	// BillingPeriod see : https://schema.org/billingPeriod
	// The time interval used to compute the invoice.
	// types : Duration
	BillingPeriod []*Duration `json:"billingPeriod,omitempty"`

	// Broker see : https://schema.org/broker
	// An entity that arranges for an exchange between a buyer and a seller.  In most cases a broker never acquires or releases ownership of a product or service involved in an exchange.  If it is not clear whether an entity is a broker, seller, or buyer, the latter two terms are preferred. Supersedes bookingAgent (see: https://schema.org/bookingAgent).
	// types : Organization Person
	Broker []interface{} `json:"broker,omitempty"`

	// Category see : https://schema.org/category
	// A category for the item. Greater signs or slashes can be used to informally indicate a category hierarchy.
	// types : PhysicalActivityCategory Text Thing
	Category []interface{} `json:"category,omitempty"`

	// ConfirmationNumber see : https://schema.org/confirmationNumber
	// A number that confirms the given order or payment has been received.
	// types : Text
	ConfirmationNumber []string `json:"confirmationNumber,omitempty"`

	// Customer see : https://schema.org/customer
	// Party placing the order or paying the invoice.
	// types : Organization Person
	Customer []interface{} `json:"customer,omitempty"`

	// MinimumPaymentDue see : https://schema.org/minimumPaymentDue
	// The minimum payment required at this time.
	// types : MonetaryAmount PriceSpecification
	MinimumPaymentDue []interface{} `json:"minimumPaymentDue,omitempty"`

	// PaymentDueDate see : https://schema.org/paymentDueDate
	// The date that payment is due. Supersedes paymentDue (see: https://schema.org/paymentDue).
	// types : DateTime
	PaymentDueDate []DateTime `json:"paymentDueDate,omitempty"`

	// PaymentMethod see : https://schema.org/paymentMethod
	// The name of the credit card or other method of payment for the order.
	// types : PaymentMethod
	PaymentMethod []*PaymentMethod `json:"paymentMethod,omitempty"`

	// PaymentMethodId see : https://schema.org/paymentMethodId
	// An identifier for the method of payment used (e.g. the last 4 digits of the credit card).
	// types : Text
	PaymentMethodId []string `json:"paymentMethodId,omitempty"`

	// PaymentStatus see : https://schema.org/paymentStatus
	// The status of payment; whether the invoice has been paid or not.
	// types : PaymentStatusType Text
	PaymentStatus []interface{} `json:"paymentStatus,omitempty"`

	// Provider see : https://schema.org/provider
	// The service provider, service operator, or service performer; the goods producer. Another party (a seller) may offer those services or goods on behalf of the provider. A provider may also serve as the seller. Supersedes carrier (see: https://schema.org/carrier).
	// types : Organization Person
	Provider []interface{} `json:"provider,omitempty"`

	// ReferencesOrder see : https://schema.org/referencesOrder
	// The Order(s) related to this Invoice. One or more Orders may be combined into a single Invoice.
	// types : Order
	ReferencesOrder []*Order `json:"referencesOrder,omitempty"`

	// ScheduledPaymentDate see : https://schema.org/scheduledPaymentDate
	// The date the invoice is scheduled to be paid.
	// types : Date
	ScheduledPaymentDate []Date `json:"scheduledPaymentDate,omitempty"`

	// TotalPaymentDue see : https://schema.org/totalPaymentDue
	// The total amount due.
	// types : MonetaryAmount PriceSpecification
	TotalPaymentDue []interface{} `json:"totalPaymentDue,omitempty"`
}

func (v Invoice) IntoMap(intop *map[string]interface{}) error {
	if intop == nil {
		return nil
	}

	v.Intangible.IntoMap(intop)

	into := *intop

	{
		var value interface{} = v.AccountId
		if len(v.AccountId) == 1 {
			value = v.AccountId[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["accountId"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.BillingPeriod
		if len(v.BillingPeriod) == 1 {
			value = v.BillingPeriod[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["billingPeriod"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.Broker
		if len(v.Broker) == 1 {
			value = v.Broker[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["broker"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.Category
		if len(v.Category) == 1 {
			value = v.Category[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["category"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ConfirmationNumber
		if len(v.ConfirmationNumber) == 1 {
			value = v.ConfirmationNumber[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["confirmationNumber"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.Customer
		if len(v.Customer) == 1 {
			value = v.Customer[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["customer"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.MinimumPaymentDue
		if len(v.MinimumPaymentDue) == 1 {
			value = v.MinimumPaymentDue[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["minimumPaymentDue"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.PaymentDueDate
		if len(v.PaymentDueDate) == 1 {
			value = v.PaymentDueDate[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["paymentDueDate"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.PaymentMethod
		if len(v.PaymentMethod) == 1 {
			value = v.PaymentMethod[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["paymentMethod"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.PaymentMethodId
		if len(v.PaymentMethodId) == 1 {
			value = v.PaymentMethodId[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["paymentMethodId"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.PaymentStatus
		if len(v.PaymentStatus) == 1 {
			value = v.PaymentStatus[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["paymentStatus"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.Provider
		if len(v.Provider) == 1 {
			value = v.Provider[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["provider"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ReferencesOrder
		if len(v.ReferencesOrder) == 1 {
			value = v.ReferencesOrder[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["referencesOrder"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ScheduledPaymentDate
		if len(v.ScheduledPaymentDate) == 1 {
			value = v.ScheduledPaymentDate[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["scheduledPaymentDate"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.TotalPaymentDue
		if len(v.TotalPaymentDue) == 1 {
			value = v.TotalPaymentDue[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["totalPaymentDue"] = json.RawMessage(b)
		}
	}

	*intop = into

	return nil
}

func (v Invoice) AsMap() (map[string]interface{}, error) {
	data := map[string]interface{}{}
	err := v.IntoMap(&data)
	if err != nil {
		return nil, err
	}

	data["@context"] = "http://schema.org"
	data["@type"] = "Invoice"

	return data, nil
}

func (v Invoice) MarshalJSON() ([]byte, error) {
	data, err := v.AsMap()
	if err != nil {
		return nil, err
	}

	return json.Marshal(data)
}

File: schemaorgphp/classes/invoice.php
<?php

class Invoice extends Intangible implements JsonSerializable {

	public static $context = 'http://schema.org';
	public static $type = 'Invoice';
	
	/**
	 * The identifier for the account the payment will be applied to.
	 * see : https://schema.org/accountId
	 * @var string|string[]
	 */
	public var $account_id;
	
	/**
	 * The time interval used to compute the invoice.
	 * see : https://schema.org/billingPeriod
	 * @var \Duration|\Duration[]
	 */
	public var $billing_period;
	
	/**
	 * An entity that arranges for an exchange between a buyer and a seller.  In most cases a broker never acquires or releases ownership of a product or service involved in an exchange.  If it is not clear whether an entity is a broker, seller, or buyer, the latter two terms are preferred. Supersedes bookingAgent (see: https://schema.org/bookingAgent).
	 * see : https://schema.org/broker
	 * @var \Organization|\Organization[]|\Person|\Person[]
	 */
	public var $broker;
	
	/**
	 * A category for the item. Greater signs or slashes can be used to informally indicate a category hierarchy.
	 * see : https://pending.schema.org/category
	 * @var \PhysicalActivityCategory|\PhysicalActivityCategory[]|string|string[]|\Thing|\Thing[]
	 */
	public var $category;
	
	/**
	 * A number that confirms the given order or payment has been received.
	 * see : https://schema.org/confirmationNumber
	 * @var string|string[]
	 */
	public var $confirmation_number;
	
	/**
	 * Party placing the order or paying the invoice.
	 * see : https://schema.org/customer
	 * @var \Organization|\Organization[]|\Person|\Person[]
	 */
	public var $customer;
	
	/**
	 * The minimum payment required at this time.
	 * see : https://schema.org/minimumPaymentDue
	 * @var \MonetaryAmount|\MonetaryAmount[]|\PriceSpecification|\PriceSpecification[]
	 */
	public var $minimum_payment_due;
	
	/**
	 * The date that payment is due. Supersedes paymentDue (see: https://schema.org/paymentDue).
	 * see : https://schema.org/paymentDueDate
	 * @var string|string[]
	 */
	public var $payment_due_date;
	
	/**
	 * The name of the credit card or other method of payment for the order.
	 * see : https://schema.org/paymentMethod
	 * @var \PaymentMethod|\PaymentMethod[]
	 */
	public var $payment_method;
	
	/**
	 * An identifier for the method of payment used (e.g. the last 4 digits of the credit card).
	 * see : https://schema.org/paymentMethodId
	 * @var string|string[]
	 */
	public var $payment_method_id;
	
	/**
	 * The status of payment; whether the invoice has been paid or not.
	 * see : https://schema.org/paymentStatus
	 * @var \PaymentStatusType|\PaymentStatusType[]|string|string[]
	 */
	public var $payment_status;
	
	/**
	 * The service provider, service operator, or service performer; the goods producer. Another party (a seller) may offer those services or goods on behalf of the provider. A provider may also serve as the seller. Supersedes carrier (see: https://schema.org/carrier).
	 * see : https://schema.org/provider
	 * @var \Organization|\Organization[]|\Person|\Person[]
	 */
	public var $provider;
	
	/**
	 * The Order(s) related to this Invoice. One or more Orders may be combined into a single Invoice.
	 * see : https://schema.org/referencesOrder
	 * @var \Order|\Order[]
	 */
	public var $references_order;
	
	/**
	 * The date the invoice is scheduled to be paid.
	 * see : https://schema.org/scheduledPaymentDate
	 * @var string|string[]
	 */
	public var $scheduled_payment_date;
	
	/**
	 * The total amount due.
	 * see : https://schema.org/totalPaymentDue
	 * @var \MonetaryAmount|\MonetaryAmount[]|\PriceSpecification|\PriceSpecification[]
	 */
	public var $total_payment_due;
	
	public function jsonSerialize() {
		$out = array(
			'@context' => 'http://schema.org',
			'@type' => 'Invoice'
		);
		
		$serialized = so_json_serialize( $this->account_id );
		if ( ! empty( $serialized ) ) {
			$out['accountId'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->billing_period );
		if ( ! empty( $serialized ) ) {
			$out['billingPeriod'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->broker );
		if ( ! empty( $serialized ) ) {
			$out['broker'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->category );
		if ( ! empty( $serialized ) ) {
			$out['category'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->confirmation_number );
		if ( ! empty( $serialized ) ) {
			$out['confirmationNumber'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->customer );
		if ( ! empty( $serialized ) ) {
			$out['customer'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->minimum_payment_due );
		if ( ! empty( $serialized ) ) {
			$out['minimumPaymentDue'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->payment_due_date );
		if ( ! empty( $serialized ) ) {
			$out['paymentDueDate'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->payment_method );
		if ( ! empty( $serialized ) ) {
			$out['paymentMethod'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->payment_method_id );
		if ( ! empty( $serialized ) ) {
			$out['paymentMethodId'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->payment_status );
		if ( ! empty( $serialized ) ) {
			$out['paymentStatus'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->provider );
		if ( ! empty( $serialized ) ) {
			$out['provider'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->references_order );
		if ( ! empty( $serialized ) ) {
			$out['referencesOrder'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->scheduled_payment_date );
		if ( ! empty( $serialized ) ) {
			$out['scheduledPaymentDate'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->total_payment_due );
		if ( ! empty( $serialized ) ) {
			$out['totalPaymentDue'] = $serialized;
		}
		
		return $out;
	}
}
